ADAL/src/ADAuthorityValidation.m

```diff
@@ -0,0 +1,396 @@
+            // Error response from the server
+            errorDetails = errorDetails ? errorDetails : [NSString stringWithFormat:@"%@ - %@", s_kValidationServerError, serverOAuth2Error];
+            
+            error = [ADAuthenticationError errorFromAuthenticationError:AD_ERROR_DEVELOPER_AUTHORITY_VALIDATION
+                                                           protocolCode:serverOAuth2Error
+                                                           errorDetails:errorDetails
+                                                          correlationId:_correlationId];
+        }
+        else
+        {
+            [self addValidAuthority:authority];
+        }
+        
+        completionBlock(verified, error);
+    }];
+}
+
+
+
+#pragma mark - ADFS authority validation
+- (void)validateADFSAuthority:(NSURL *)authority
+                       domain:(NSString *)domain
+              completionBlock:(void (^)(BOOL validated, ADAuthenticationError *error))completionBlock
+{
+    // Check cache first
+    if([self isAuthorityValidated:authority domain:domain])
+    {
+        completionBlock(YES, nil);
+        return;
+    }
+    
+    // DRS discovery
+    [self requestDrsDiscovery:domain
+              completionBlock:^(id result, ADAuthenticationError *error)
+    {
+        NSString *passiveAuthEndpoint = [self passiveEndpointFromDRSMetaData:result];
+
+        if (!passiveAuthEndpoint)
+        {
+            if (!error)
+            {
+                error = [ADAuthenticationError errorFromAuthenticationError:AD_ERROR_DEVELOPER_AUTHORITY_VALIDATION
+                                                               protocolCode:nil
+                                                               errorDetails:s_kDrsDiscoveryError
+                                                              correlationId:_correlationId];
+            }
+            completionBlock(NO, error);
+            return;
+        }
+        
+        [self requestWebFingerValidation:passiveAuthEndpoint
+                               authority:authority
+                         completionBlock:^(BOOL validated, ADAuthenticationError *error)
+        {
+            if (validated)
+            {
+                [self addValidAuthority:authority domain:domain];
+            }
+            completionBlock(validated, error);
+        }];
+    }];
+}
+
+- (void)requestDrsDiscovery:(NSString *)domain
+            completionBlock:(void (^)(id result, ADAuthenticationError *error))completionBlock
+{
+    [ADDrsDiscoveryRequest requestDrsDiscoveryForDomain:domain
+                                               adfsType:AD_ADFS_ON_PREMS
+                                                context:self
+                                        completionBlock:^(id result, ADAuthenticationError *error)
+    {
+        if (!result)
+        {
+            [ADDrsDiscoveryRequest requestDrsDiscoveryForDomain:domain
```

Comment:
When you start nesting blocks like this I usually prefer to just see an early return at the top. Especially if the code starts expanding it can get more difficult to trace the execution flow. e.g.:

```
if (result)
{
    completionBlock(result, nil);
    return;
}

// Do more stuff
```